### PR TITLE
allow adding private repositories using access-tokens

### PR DESCRIPTION
**Pull request by c0rydoras**

### CODE REVIEW

File: api/outdated/outdated/tests/test_validators.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+
+    serializer = ProjectSerializer(project)
+
+    with suppress(ValidationError):
+        validators.validate_remote_url(attrs, serializer)
+
+    if project.repo == repo and project.repo_type == repo_type:
+        check_remote_existance_mock.assert_not_called()
+    else:
+        check_remote_existance_mock.assert_called()
```
Comment (open-dynaMIX) [inline api/outdated/outdated/tests/test_validators.py:157]: Not sure if I understand this 100% right.

```python
    try:
        validators.validate_remote_url(attrs, serializer)
    except ValidationError:
        assert project.repo == repo and project.repo_type == repo_type
        check_remote_existance_mock.assert_not_called()
    else:
        assert project.repo != repo and project.repo_type != repo_type
        check_remote_existance_mock.assert_called()
```

File: api/outdated/outdated/migrations/0001_initial.py [hunk 2]
```diff
@@ -1,4 +1,4 @@
-# Generated by Django 4.2.6 on 2023-10-24 14:19
+# Generated by Django 4.2.6 on 2024-01-09 09:41
```
Comment (open-dynaMIX) [inline api/outdated/outdated/migrations/0001_initial.py:1]: I would prefer having migrations formatted too.

**Review by open-dynaMIX — CHANGES_REQUESTED** [review]
I only reviewed the backend part.

Awesome work :rocket:

Still needs some cleanup. I also have some further questions regarding `api/outdated/outdated/tests/test_validators.py` we might better discuss in person. Please reach out to me, when you find the time.

File: api/outdated/outdated/serializers.py [hunk 3]
```diff
@@ -50,30 +57,66 @@ class ProjectSerializer(serializers.ModelSerializer):
...
             "versioned_dependencies",
             "maintainers",
         )
 
-    def create(self, validated_data):
+    def create(self, validated_data: dict) -> models.Project:
+        access_token = None
+        if "access_token" in validated_data:
+            access_token = validated_data["access_token"]
+            del validated_data["access_token"]
```
[the remark this answers is not in the document]
Comment (c0rydoras) [reply api/outdated/outdated/serializers.py]: done here: https://github.com/adfinis/Outdated/pull/178/commits/462681b9a0b7f157620eab48004b3184b14c076d

File: api/outdated/outdated/serializers.py [hunk 4]
```diff
@@ -50,30 +57,66 @@ class ProjectSerializer(serializers.ModelSerializer):
...
+        Tracker(instance, access_token).setup()
         return instance
 
     def update(self, instance: models.Project, validated_data: dict) -> models.Project:
-        old_instance = models.Project(repo=instance.repo)
+        old_instance = models.Project(repo=instance.repo, repo_type=instance.repo_type)
+        access_token = None
+        if "access_token" in validated_data:
+            access_token = validated_data["access_token"]
+            del validated_data["access_token"]
```
[the remark this answers is not in the document]
Comment (c0rydoras) [reply api/outdated/outdated/serializers.py]: done here: https://github.com/adfinis/Outdated/pull/178/commits/462681b9a0b7f157620eab48004b3184b14c076d

File: api/outdated/outdated/validators.py [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+from outdated.validators import with_context
+
+if TYPE_CHECKING:
+    from typing import Any
+
+    from outdated.outdated.serializers import ProjectSerializer
+
+
+def check_remote_existance(url: str) -> bool:
```
[the remark this answers is not in the document]
Comment (c0rydoras) [reply api/outdated/outdated/validators.py]: fixed here: https://github.com/adfinis/Outdated/pull/178/commits/df1dc42650a91ff88d828af649ad0a04410f7100

File: api/outdated/outdated/tests/test_tracking.py [hunk 6]
```diff
@@ -12,13 +12,14 @@
...
-        ("github.com/adfinis/outdated", False),
-        ("Github.Com/ADFINIS/OutdAted", False),
-        ("github.com/adfinis/mysagw", True),
-        ("github.com/adfinis/timed-frontend", True),
+        ("github.com/Adfinis/Outdated", "public", None, False),
+        ("github.com/adfinis/outdated", "access-token", "token", False),
+        ("Github.Com/ADFINIS/OutdAted", "public", None, False),
+        ("github.com/adfinis/mysagw", "public", None, True),
+        ("github.com/adfinis/timed-frontend", "public", None, True),
+        ("github.com/adfinis/timed-backend", "public", "token", True),
```
[the remark this answers is not in the document]
Comment (c0rydoras) [reply api/outdated/outdated/tests/test_tracking.py]: Yes it should, im not sure why i even had access_token in that test, i changed it here: https://github.com/adfinis/Outdated/pull/178/commits/13aabef233dba849ee7f422853d5e8ca2c2de70c

File: api/outdated/outdated/migrations/0001_initial.py (hunk 2 above)
Comment (c0rydoras) [reply api/outdated/outdated/migrations/0001_initial.py:1]: done here: https://github.com/adfinis/Outdated/pull/178/commits/b922e7d1002fcd1909fee425bd780a7a7dac40e8

File: api/outdated/outdated/tests/test_validators.py [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+    serializer = ProjectSerializer(project if instance else None)
+
+    try:
+        validators.validate_access_token_required(attrs, serializer)
+        if repo_type == "access-token" and not access_token:
+            assert serializer.instance
+        elif repo_type == "access-token":
+            assert access_token
+        else:
+            assert repo_type == "public"
```
[the remark this answers is not in the document]
Comment (c0rydoras) [reply api/outdated/outdated/tests/test_validators.py]: i have rewritten this test

File: api/outdated/outdated/tests/test_validators.py (hunk 1 above)
Comment (c0rydoras) [reply api/outdated/outdated/tests/test_validators.py]: i have rewritten this test

File: api/outdated/outdated/tests/test_validators.py [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+    }
+
+    serializer = ProjectSerializer(project if instance else None)
+
+    try:
+        validators.validate_access_token_required(attrs, serializer)
+        if repo_type == "access-token" and not access_token:
+            assert serializer.instance
+        elif repo_type == "access-token":
+            assert access_token
```
[the remark this answers is not in the document]
Comment (c0rydoras) [reply api/outdated/outdated/tests/test_validators.py:51]: i rewrote this test

File: api/outdated/outdated/tracking.py [hunk 9]
```diff
@@ -103,7 +109,7 @@ def setup(self):  # pragma: no cover
         self.checkout()
         self.sync()
 
-    def delete(self):
+    def delete(self):  # pragma: no cover
```
[the remark this answers is not in the document]
Comment (c0rydoras) [reply api/outdated/outdated/tracking.py:112]: done here: https://github.com/adfinis/Outdated/pull/178/commits/a8f8f057c2a0318d97f0f7de5796b7b30af15953

File: api/outdated/outdated/tests/test_validators.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (open-dynaMIX) [reply api/outdated/outdated/tests/test_validators.py:51]: Great work! Would be nice if you could add some comments to the test, especially inside the try/except block. Took some time to see what exactly is done here and some explanation would have helped.

[the remark this answers is not in the document]
Comment (open-dynaMIX) [reply api/outdated/outdated/tests/test_validators.py:51]: I'm gonna approve anyway. Feel free to still add some comments :)

**Review by open-dynaMIX — APPROVED** [review]

File: ember/app/components/form.hbs [hunk 10]
```diff
@@ -8,7 +8,10 @@
         model=@model
         loading=this.loading
         input=(component
-          "validated-input" model=@model submitted=this.submitted
+          "validated-input"
+          model=@model
+          submitted=this.submitted
+          validateModel=this.validateModel
```
Comment (anehx) [inline ember/app/components/form.hbs:14]: I don't see `this.validateModel` being defined anywhere.

Comment (c0rydoras) [reply ember/app/components/form.hbs:14]: `this.validateModel` was lost when rebasing after #170 was merged (exact commit: https://github.com/adfinis/Outdated/pull/170/commits/51940586953e21bc04cbb6f021a7813a4cf13213)

I have now re implemented this:
https://github.com/adfinis/Outdated/blob/963398af242e6dccb65d97490ccd189a720b7ee9/ember/app/components/form.js#L10-L21
https://github.com/adfinis/Outdated/blob/963398af242e6dccb65d97490ccd189a720b7ee9/ember/app/components/validated-input/render/component.js#L8-L14

**Review by anehx — APPROVED** [review]

**Pull request merged**